src/sender.js

```diff
@@ -157,53 +330,87 @@ class Sender {
                 }
             })
             .on('ready', async () => {
-                this.log('info', `Successfully connected to ${options.host}:${options.port}`);
+                this.log('info', `Successfully connected to ${connectOptions.host}:${connectOptions.port}`);
                 if (self.jwk) {
-                    this.log('info', `Authenticating with ${options.host}:${options.port}`);
+                    this.log('info', `Authenticating with ${connectOptions.host}:${connectOptions.port}`);
                     await self.socket.write(`${self.jwk.kid}\n`, err => {
                         if (err) {
                             reject(err);
                         }
                     });
                 } else {
                     authenticated = true;
-                    resolve(true);
+                    if (!self.secure || !self.tlsVerify) {
+                        resolve(true);
+                    }
                 }
             })
             .on('error', err => {
-                this.log('error', err);
-                reject(err);
+                self.log('error', err);
+                if (err.code !== 'SELF_SIGNED_CERT_IN_CHAIN' || self.tlsVerify) {
+                    reject(err);
+                }
             });
         });
     }
 
     /**
-     * Closes the connection to the database. <br>
+     * @ignore
+     * @return {http.Agent} Returns the default http agent.
+     */
+    getDefaultHttpAgent() {
+        if (!Sender.DEFAULT_HTTP_AGENT) {
+            Sender.DEFAULT_HTTP_AGENT = new http.Agent(DEFAULT_HTTP_AGENT_CONFIG);
+        }
+        return Sender.DEFAULT_HTTP_AGENT;
+    }
+
+    /**
+     * @ignore
+     * @return {https.Agent} Returns the default https agent.
+     */
+    getDefaultHttpsAgent() {
+        if (!Sender.DEFAULT_HTTPS_AGENT) {
+            Sender.DEFAULT_HTTPS_AGENT = new https.Agent(DEFAULT_HTTP_AGENT_CONFIG);
+        }
+        return Sender.DEFAULT_HTTPS_AGENT;
+    }
+
+    /**
+     * Closes the TCP connection to the database. <br>
      * Data sitting in the Sender's buffer will be lost unless flush() is called before close().
      */
     async close() {
-        const address = this.socket.remoteAddress;
-        const port = this.socket.remotePort;
-        this.socket.destroy();
-        this.log('info', `Connection to ${address}:${port} is closed`);
+        if (this.socket) {
+            const address = this.socket.remoteAddress;
+            const port = this.socket.remotePort;
+            this.socket.destroy();
```

Comment:
nit: we should probably set `this.socket` to null too.

Reply:
changed it